### ISSUE TITLE
Discussion: X# / Xaml Sharp / Xaml Code - inspired by SwiftUI

### ISSUE THREAD

**Issue by mdtauk** [microsoft/microsoft-ui-xaml#804]
Apple announced SwiftUI which replaces an older and much more verbose method of building UI for their iOS, iPadOS and macOS platforms.

![image](https://user-images.githubusercontent.com/7389110/58920455-30e1ee80-872a-11e9-95c5-630b061f65d6.png)

This is not something that could be done by time of WinUI 3.0 and so is not a proposal per se, but XAML being XML based, is nothing if not verbose.  A lot of effort has been made in recent years to make it possible to write less Xaml to do more.  VisualStateTriggers being a massive one replacing the VisualStateManager.  There is also a proposal to re-work the Grid panel control, to make it more powerful with less Xaml required.

XamlDirect currently exists aimed at middleware tools which can generate XAML from code, but is there some merit into looking at what SwiftUI is doing to perhaps create an alternative to Xaml Markup, with the same power and flexibility of Xaml, but similar to React Native, providing a less verbose version of Xaml.

What it could be called is anyone's guess - for the sake of this suggestion I call it X#, XamlSharp (perhaps inspired by C#'s use of the dot property syntax)
```
using xaml;  
  
xaml page  
{  
      this.backgroundColor = themeResource(ApplicationPageBackgroundColor);  
      this.foregroundColor = themeResource(ApplicationTextForegroundColor);  
  
      xaml grid[name "rootGrid", rows "*, *, *, Auto, 12epx"]  
      {  
            xaml stackPanel[name "toolbar", orientation Horizontal]  
            {}  
            xaml button[grid.row "4" content xaml bind(_appIcon)];  
      }  
}  
```
^ Example code is not a suggested syntax, but just an illustrative example

### This is a place where anyone who has an interest in simplifying Xaml can share their thoughts, ideas, syntax suggestions

**Comment by RichiCoder1** [comment]
Adding Ammy UI (http://www.ammyui.com/) as excellent prior art in this area.

**Comment by mdtauk** [comment]
Ammy is a good attempt at making XAML UI simpler, but it does lean too much towards CSS than more familiar code based approaches that may be more familiar to Microsoft developers, and would work more seamlessly with current tooling I would expect.

JSON is very popular at the moment, but it is so fickle and easy to break - where as XML is robust but complex and wordy.

Whatever format is chosen, there should be a thought on constructing the UI on the screen decoratively, but an ability to easily manage Templates and Styles either inline or as attached files.  At the moment XAML does it all, but it is bothersome to edit for designers, and also verbose for those who prefer to work in code.

**Comment by contextfree** [comment]
It's already possible to make WPF or WinUI UI purely in code (for example, Charles Petzold's WPF book didn't even introduce XAML at all until halfway through the book) so it might be good to use this issue to call out specific problems with doing this today (bindings and templates for example?)

It also might be worth looking at [Fabulous](https://github.com/fsprojects/Fabulous) which is an existing project for F# and WPF or Xamarin.Forms that seems along the lines suggested in this topic.

**Comment by mdtauk** [comment]
> It's already possible to make WPF or WinUI UI purely in code (for example, Charles Petzold's WPF book didn't even introduce XAML at all until halfway through the book) so it might be good to use this issue to call out specific problems with doing this today (bindings and templates for example?)
> 
The thing that makes SwiftUI stand out is that it remains declarative in nature, similar to HTML and Markup, but is within the language and syntax of Swift.  It is possible to create XAML objects, but the concept of nesting elements, and declaring the properties, bindings, templates and styles are obtuse and not very effective.

**Comment by Felix-Dev** [comment]
I guess it will come down to available resources at Microsoft. For the forseeable future, I want them to first-and-foremost focus on making XAML less verbose - and there have been countless of proposals over the years how to do just that. WinUI 3.0 is currently blocking most of that work so we need to wait some more months until MS and the community can work together and give XAML a much needed update across the board. Depending on how that will go, we can re-visit your idea, Martin, and potentially look at areas outside of XAML markup to make UI design much more efficient.

**Comment by Happypig375** [comment]
@mdtauk
That looks awfully similar to what we can do today:
```cs
using System;
using Windows.ApplicationModel.Activation;
using Windows.UI;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;
using Windows.UI.Xaml.Media;
namespace NoXaml
{
    static class Program { public static void Main(string[] _) => Application.Start(p => new MyApp()); }
    class MyApp : Application
    {
        protected override void OnLaunched(LaunchActivatedEventArgs args)
        {
            Window.Current.Content = MainPage;
            Window.Current.Activate();
        }

        UIElement MainPage =>
            new Grid
            {
                Background = new SolidColorBrush(Colors.Blue),
                Children = {
                    new Button { Content = "Hello!" }
                }
            };
    }
}
```

**Comment by Mike-EEE** [comment]
Adding @VincentH-Net's CSharpForMarkup to the mix as well for reference. https://github.com/VincentH-Net/CSharpForMarkup

FWIW, ATM I think this is a terrible idea but I am keeping an open mind about it.  Maybe it will grow on me. 😁

> Whatever format is chosen

Ideally, you should be able to choose between the two (or more) formats, for optimal adoption.  This is the forward-thinking and .NET-like approach that @cwensley utilizes for his ETO project: https://github.com/picoe/Eto

**Comment by dpierangeli** [comment]
Ok, it can be an useful addition, but please don't do this:

![swiftuiblah](https://user-images.githubusercontent.com/9047951/58943785-ca3bef80-8780-11e9-950d-41006091ea5b.jpg)

A declarative UI definition should not be mixed with application logic (my idea).

**Comment by Tirraon** [comment]
Avalonia team also discussed about XAML markup alternative:
https://github.com/AvaloniaUI/Avalonia/issues/2502

**Comment by TonyHenrique** [comment]
XAML markup is a advancement. These approaches are limited. .NET and XAML need to move forward and reach those platforms, not the contrary.
XAML is a way to instantiate objects and separate 'UI code' from the 'code behind'.

Also give a look at https://github.com/microsoft/microsoft-ui-xaml/issues/740#issuecomment-497946126

**Comment by TonyHenrique** [comment]
If we look carefully we see clearly the simplicity and advantages of XAML.
If XAML runs everywhere, it could be used instead of HTML in lots of use cases.
We need to be careful to don't waste time with these pseudo improvements, as they may up ending wanting to describe UI using YAML files...

**Comment by weitzhandler** [comment]
@mdtauk
My 2 cents: if we're to drop XAML, it's for the favor of a C#-only Dart-like UI. With RxUI this could work nicely.
Anyway, I'm happy with XAML the way it is, they should just add plenty of more built-in converters, interactivity actions and markup extensions, an integrated way of C# expressions, and a lot of syntax-sugars to make things less-verbose.

**Comment by Felix-Dev** [comment]
There are definitely a lot of possible XAML Markup improvements the team should focus on before thinking about alternative ways to build a UI. Both on the control level (like the new grid proposal) and the XAML language itself like introducing a negation syntax for binding, i.e: 
`IsVisible = "{x:Bind !SomeObject.SomeProperty}"`
or stuff mentioned in these proposals: https://github.com/microsoft/microsoft-ui-xaml/issues/62 and https://github.com/dotnet/wpf/issues/480 (and there are a ton more)

To sum it up: The focus should be first-and-foremost on improving XAML markup!

**Comment by mdtauk** [comment]
Offering the choice of building XAML UI in C#, C++, F#, VB, JSON, as well as Xaml markup itself - would allow the developer to choose what way is best for them - as long as it all renders the same.

There is Xaml Direct **Microsoft.UI.Xaml.Core.Direct** but it is intended for middle-ware tools, is very low level, and hasn't been formalised as a first class way to build an app's UI. (New Project templates, sample code, etc)

I am not averse to Xaml as Markup (but anything that results in less markup and simpler declarations would be welcome) but SwiftUI shows that there are alternatives.  No idea if one is better than the other, but it is interesting to discuss the pain points, the aspirations, and to possibly learn lessons from other platform ideas :)

**Comment by RyoukoKonpaku** [comment]
I do agree with the rest that XAML improvements are of priority first and foremost. 

For the sake of discussion on possible alternatives how about talking with the Razor Compiler team? They seem to have a decent compiler already that can parse C# and VB (not too sure if this is still supported on newer razor) expressions in `HTML` and instantiating C# classes (Blazor Components in this case). It's `HTML` is actually more like `XML` as well since Razor is kinda strict with closing tags for their compiler which makes it a good candidate for an alternative. Plus points since it's also an MS product so they can freely communicate between the two if necessary.

Off the top of my head it'd probably be like this? This is assuming we treat `Xaml` classes as Blazor components so that there's less effort needed to make it work.
```
@* .csxaml *@
<Grid>
    @* Using the new directive support that .razor has *@
    <Grid.Row @Bind:Height="someHeightVariable"
                      @Bind:Mode="Binding.TwoWay">
         <Grid.Column>
             @* Can use lambda or for MVVM be a method on the ViewModel *@
             <Button PlaceHolder="Todo" 
                           @Command="ViewModel.SomeMethod"
                           @Background="ConstantsClass.SomeBrush">
                   Some Button
             </Button>
         </Grid.Column>
    </Grid.Row>
</Grid>
```

This should be less effort in my opinion that writing a new DSL parser from scratch, but the downside is the razor compiler is biased to C# by default so adding other languages like F# might be a hassle.

**Comment by Mike-EEE** [comment]
I like your thinking @RyoukoKonpaku.  Unfortunately, Razor is a templating engine, and the results it generates is a plain text file (or stream) for use to emit to a browser.  What we're attempting to do here is describe an object graph that ultimately gets loaded into memory and rendered from there.

For instance, while you do reference a `Grid` object, there is no way for Razor to know what exactly that is, so there is no way to load it into memory for further processing.

Unless, of course, you're suggesting that we use Razor to create XAML files and _then_ from there we load in the Xaml, which would work.  In that case, however, you're doubling the required components to get the view loaded, plus there are other considerations such as debugging.  How do you debug a binding that was defined in a template file that created the Xaml file?

Not to say that it wouldn't work.  It's a neat idea, but it would be an involved undertaking, IMO.

**Comment by mdtauk** [comment]
One point I think most of us would agree on, is that whatever emerges should be **Human Readable**

**Comment by TonyHenrique** [comment]
I would add that I feel that if we unite and ask for the XAML Standard or a Universal XAML, it would be much better that any attempt to change or improve the XAML.

About the code only approach, the best to do would be invest in F# elmish style and also gjallahorn with a WinUI XAML Type Provider.

Another thing that could be done is to allow XAML literals in C#/F# code. 
This would enable code like this

let myButton title = <Button Text=title />

**Comment by RyoukoKonpaku** [comment]
@Mike-EEE Indeed it's not that smooth sailing either. I'm actually referring to Blazor in this case which can detect C# classes as `Components` in Blazor terms. The `.razor` file actually compiles into a render instruction set on a class. So I'd assume it's possible to change the generated class from the AST that Razor parses.

```html
@* SampleComponent.razor *@
<ChildComponent>
    <NestedComponent />
</ChildComponent>
```

The generated class is this.
```cs
// compiled SampleComponent.g.cs
# alot of pragmas here since it's generated
public partial class SampleComponent : Microsoft.AspNetCore.Components.ComponentBase
{
    protected override void BuildRenderTree(Microsoft.AspNetCore.Components.RenderTree.RenderTreeBuilder builder) {
        builder.OpenComponent<Namespace.ChildComponent>();
        builder.OpenComponent<Namespace.NestedComponent>();
        builder.CloseComponent<Namespace.NestedComponent>();
        builder.CloseComponent<Namespace.ChildComponent>();
        // ... removed for brevity but basically constructs the combination of 
        // strings (e.g. div, span etc...) and Component which are eseentially just classes
    }
}
```

So what I'd imagine is that we replace the generated builder into a series of `Xaml` class instantiations into an object graph via `Xaml.Direct` middleware. Also remove the ability to use html tags so it does not compile for that.

So from the above it would compile to something like this
```cs
// SampleComponent.g.cs
# pragmas here
public partial class SampleComponent: Microsoft.UI.Xaml.UserControl
{
    protected SampleComponent() {
         InitializeControl();
         // Create object graph here for ChildComponent and it's nested Control.
    }
}
```

**Comment by Mike-EEE** [comment]
To be sure, @RyoukoKonpaku, are `ChildComponent` and `NestedComponent` in  `NameSpace.SomeControls` or `Namespace` ?  My Razor/Blazor is rather newb, so I am trying to understand how it is resolving those.

In either case, that does look interesting and worth an exploration, IMO. 👍

**Comment by reli-msft** [comment]
I think this could be discussed in https://github.com/dotnet/csharplang?
It is all about extending C# to provide ability to define embedded DSLs, then you can have a UI DSL and other DSL if you need.

**Comment by RyoukoKonpaku** [comment]
@Mike-EEE yep `ChildComponent` and `NestedComponent` are C# classes that derives from `ComponentBase` on the `NameSpace.SomeControl` namespace, I kinda flunked the example abit there but it should have referenced `NameSpace.SomeControls.ChildComponent` on the generated class instead of `NameSpace.ChildComponent`.

You can also refer the full namespace to reference a Component just like how you'd reference a normal class.
```html
<NameSpace.SomeControls.ChildComponent>
    <NameSpace.SomeControls.NestedComponent />
</NameSpace.SomeControls.ChildComponent>
```

The AST to C# class is the bigger effort if this was to be done from what I can tell but it's definitely plausible with the tools available now, `Xaml.Direct`, `Razor's AST`, and knowledge on gluing those together, razor bits from the Razor compiler team and `Xaml.Direct` usage from the WinUI team.

**Comment by Mike-EEE** [comment]
You might be onto something @RyoukoKonpaku.  I'm taking @reli-msft's hint and moving this onto a perhaps more pertinent forum: https://github.com/aspnet/AspNetCore/issues/10911

See you there. :)

**Comment by bschoepke** [comment]
I would push for TypeScript instead of C# personally. It would lower the barrier to entry significantly.

**Comment by mdtauk** [comment]
> I would push for TypeScript instead of C# personally. It would lower the barrier to entry significantly.
> 
> I would also completely kill markup; the simplicity of Swift UI is that there is only a single, expressive programming language for everything. No XML or JSON to deal with.
> 
> If we do something new, you should be able to incrementally adopt it into existing apps.
> 
> The real advancement would be to enable mixing/matching web-based and native controls in the same app or PWA.

Most UI examples I have come across involving JavaScript - still relies on markup languages like XML or HTML/CSS.

**Comment by bschoepke** [comment]
> > I would push for TypeScript instead of C# personally. It would lower the barrier to entry significantly.
> > I would also completely kill markup; the simplicity of Swift UI is that there is only a single, expressive programming language for everything. No XML or JSON to deal with.
> > If we do something new, you should be able to incrementally adopt it into existing apps.
> > The real advancement would be to enable mixing/matching web-based and native controls in the same app or PWA.
> 
> Most UI examples I have come across involving JavaScript - still relies on markup languages like XML or HTML/CSS.

Some of the newer MSFT web frameworks are pure TypeScript (https://github.com/Microsoft/fast-dna) and we've found massive advantages to this. It's just much simpler to only deal with one, expressive programming language vs. a bunch of domain-specific languages.

Reminds me of the old Code == Data debates that LISP advocates like to make.

**Comment by mdtauk** [comment]
> Some of the newer MSFT web frameworks are pure TypeScript (https://github.com/Microsoft/fast-dna) and we've found massive advantages to this. It's just much simpler to only deal with one, expressive programming language vs. a bunch of domain-specific languages.
> 
> Reminds me of the old Code == Data debates that LISP advocates like to make.

That example  https://explore.fastdna.net/components/action-toggle/
Is still rendering to HTML.  For that there is already Fluent Web and Fabric Web.

I am talking about rendering the same XAML UI, but possible alternatives to the Markup, which can be very verbose and imposing with its XML style syntax and "obtuse" approach to control styling.

SwiftUI uses the same Swift code syntax, but allows XAML style concepts such as property binding, and including UI within other UI that is spread across multiple files.

This is what inspired me to open this discussion.  XAML has been around for almost 11 years, and since then:
- Android uses a declarative markup language, and Kotlin is their main code language.
- HTML/CSS has become more powerful with JavaScript coding.
- Apple has moved away from its XIB (XML Interface Builder) approach and moved to Swift.

Swift and Kotlin as with C# were designed to feel familiar to Java concepts, and are a lot more alike.

Are there lessons that can be learned?

**Comment by bschoepke** [comment]
> > Some of the newer MSFT web frameworks are pure TypeScript (https://github.com/Microsoft/fast-dna) and we've found massive advantages to this. It's just much simpler to only deal with one, expressive programming language vs. a bunch of domain-specific languages.
> > Reminds me of the old Code == Data debates that LISP advocates like to make.
> 
> That example https://explore.fastdna.net/components/action-toggle/
> Is still rendering to HTML. For that there is already Fluent Web and Fabric Web.
> 
> I am talking about rendering the same XAML UI, but possible alternatives to the Markup, which can be very verbose and imposing with its XML style syntax and "obtuse" approach to control styling.
> 
> SwiftUI uses the same Swift code syntax, but allows XAML style concepts such as property binding, and including UI within other UI that is spread across multiple files.
> 
> This is what inspired me to open this discussion. XAML has been around for almost 11 years, and since then:
> 
> * Android uses a declarative markup language, and Kotlin is their main code language.
> * HTML/CSS has become more powerful with JavaScript coding.
> * Apple has moved away from its XIB (XML Interface Builder) approach and moved to Swift.
> 
> Swift and Kotlin as with C# were designed to feel familiar to Java concepts, and are a lot more alike.
> 
> Are there lessons that can be learned?

We're on the same page. I think the major advancement of Swift UI is that there is no markup or DSL; your app is just 100% Swift.

I've been doing XAML for 10 years on multiple teams and the ramp up time is severe. Swift UI looks like it will be massively easier to start using because there are just far less concepts and DSL's to learn.

I wonder if we could just make Swift and Swift UI work on Windows via xlang and WinUI 3.0.

**Comment by michael-hawker** [comment]
[From Twitter](https://twitter.com/XAMLLlama/status/1136326473143373824)

> I always find it easier to move bits & pieces of XAML around, either for prototyping or re-use, than if I have bits of code to move around, especially subsets.
>
> Plus, it's easier to separate concerns in XAML with data binding than being tempted to smash it all together in code.

**Comment by mdtauk** [comment]
> [From Twitter](https://twitter.com/XAMLLlama/status/1136326473143373824)
> 
> > I always find it easier to move bits & pieces of XAML around, either for prototyping or re-use, than if I have bits of code to move around, especially subsets.
> > Plus, it's easier to separate concerns in XAML with data binding than being tempted to smash it all together in code.

I agree that is the good part of XAML, you can just copy the markup and move it.  But even if you used code or some other syntax - it would still be a separate file with a code behind.

**Comment by JaggerJo** [comment]
React, Flutter, Elm, Reason and now Swift UI (I think) all use a MVU like approach. 

It would be great if there was something like this for Win UI / UWP. I don't think we need yet another language, but a solid abstraction we can agree and build DSL on top.

**Comment by mdtauk** [comment]
![image](https://user-images.githubusercontent.com/7389110/59065458-a376d980-88a4-11e9-8496-6d7ff52cc007.png)

![image](https://user-images.githubusercontent.com/7389110/59065484-b4bfe600-88a4-11e9-9d71-cd39b9fa2a5e.png)

![image](https://user-images.githubusercontent.com/7389110/59065532-d620d200-88a4-11e9-9f59-31284c08c24a.png)

These are all very similar.  The C# concept of a Model is out of scope for this discussion, but we can presume that remains as is.  The only change we need focus on is the View.  The Controller, or Update step would be the code-behind portion of this, and so would remain as is - but the ViewModel seems like a clunky solution, purely because XAML wants to be kept separate, and each Xaml file has a code-behind that MVVM wants to avoid putting code into.

**Comment by weitzhandler** [comment]
Somehow related: https://github.com/dotnet/csharplang/issues/350#issuecomment-489240829

**Comment by ekral** [comment]
These are all very similar. The C# concept of a Model is out of scope for this discussion, but we can presume that remains as is. 

I think that MPU is like a MVVM with the immutable ViewModel. The difference is that in MVVM we define presentation model (ViewModel) containing data and UI behaviour separately from the domain model.
The weak point of the MVVM is the notification of updates using the PropertyChanged event. In MPU you just create a new model and that is all you must do.

**Comment by Mike-EEE** [comment]
FWIW, SwiftUI as a Flutter application:
https://medium.com/flutter-nyc/building-the-swiftui-sample-app-in-flutter-67bb4f9c571c

I am still not sold on this whole way of thinking, but consider myself subscribed to its newsletter and am interested to see how it evolves.  Lots of room for improvement, IMO.

**Comment by ryandemopoulos** [comment]
This is a great (and deep) topic, and over the last few years our team has sort of danced around it with ideas here-and-there.  We did some light explorations about 3 years ago at adopting a CSS-like syntax for styling.  More recently we've had some internal discussions about the pros/cons of XAML as a language, and how to evolve that moving forward vs. adopting other languages/syntaxes.  These sorts of "investigations" (if you want to call them that) have been mainly 1-off efforts and always done with the tacit understanding that it's pretty hard to develop and adopt new languages on existing technologies.  That's not to say it can't be done--just that it has a huge downstream cost of confusion and people having to learn a new way of doing things (with a huge documentation cost to help educate folks too).

At some point I think we should look at this more holistically; I appreciate you raising this topic up @mdtauk.  I could see us assigning a Program Manager from our team to this topic to take a pretty serious look at it.  In keeping with what I posted in #966, I'm going to remove the Feature Proposal label and keep this as an ongoing Discussion issue--the great comments in here will serve as a solid starting point for if/when we can take a more concentrated look at it.  For the time being, we're pretty heads-down on just trying to get WinUI 2.x and 3.0 planned, so this would certainly be a post-3.0 thing (and therefore spill into 2020).

Hope that makes sense.  Want to keep this active--the discussion is great.

**Comment by MarkIngramUK** [comment]
I would be happy with the ability to create a `ControlTemplate` in code to begin with. We can already achieve creation of UI controls in code, and it might be a little more verbose compared to SwiftUI's approach, but it's still possible. I would rather give code the same ability as XAML to start.

**Comment by RyoukoKonpaku** [comment]
@Mike-EEE You guys might be interested on this recent NDC conference with Steve Sanderson (Lead dev of Blazor). He managed to do what I assumed was possible on the above discussion regarding Blazor using non HTML controls (in this case native WinUI controls). [Here's the NDC link and it's around 54 min mark](https://youtu.be/uW-Kk7Qpv5U?t=3255) where he uses a Flutter backend + Blazor to render a fully cross platform native UI.

This might be a plausible path that WinUI can integrate with to enable full C# capability within the UI framework like JSX via React Native then slowly enable xplatform capabilities and you'll have access to Web, Mobile, and Desktop platforms.

**Comment by Mike-EEE** [comment]
> You guys might be interested on this recent NDC conference with Steve Sanderson (Lead dev of Blazor)

I am amazed.  Blamazed even (sorry, couldn't help it -- but now have the name for my next project 🤣).

If @stevesanderson / @danroth27 / Blazor leadership team doesn't productize this Blutter product idea (with a better name, of course 😉), it will **easily** be the worst disaster MSFT mistake since Clippy shelved Silverlight. 😆

You are basically saying you have built an idiomatic bridge to your direct competition, which allows developers to use C# rather than Dart while also (!) utilizing Visual Studio rather than whatever Flutter is using, all under the power of the Blazor API.  I am not sure why you would NOT want to productize this?!?!?!

**Comment by weitzhandler** [comment]
@Mike-EEE I've watched the whole video through and looks amazing indeed.
However, I'm still more a fan of sticking to non-HTML UIs we're already used to, such as XAML. BUT, it might be a great idea if we could integrate C# snippets into XAML like Razor does. If they do that, that'd be a complete game changer for XAML.

**Comment by ekral** [comment]
> @Mike-EEE I've watched the whole video through and looks amazing indeed.
> However, I'm still more a fan of sticking to non-HTML UIs we're already used to, such as XAML. BUT, it might be a great idea if we could integrate C# snippets into XAML like Razor does. If they do that, that'd be a complete game changer for XAML.



> @Mike-EEE I've watched the whole video through and looks amazing indeed.
> However, I'm still more a fan of sticking to non-HTML UIs we're already used to, such as XAML. 

Actually, their last demo in the video presents Blazor with the Flutter UI renderer instead of the HTML UI.

**Comment by ekral** [comment]
Personally I found out the Razor syntax for Flutter better then the Flutter Dart syntax or Swift UI.

**Comment by Mike-EEE** [comment]
> Personally I think that the Razor syntax for Flutter looks better then the Flutter Dart syntax or Swift UI.

Agreed @ekral.  Call me old-fashioned, but I _far_ prefer having markup represent UI/controls than code.  Conceptually, I like having a clear cognitive separation between "presentation" and "code."  I don't mind Blazor's binding syntax too much, but I can see how others might have a problem with it, especially with a Xaml background.  

I guess after suffering nearly a decade from a lack of true .NET client application development model, I am thirsty enough to tolerate it. 😆

**Comment by apws** [comment]
HI, somebody mentioned NeosisGUI (I traversed here from other issue related to xplat UI) and all I can say after some long reads is that it looks like silverlight native agcore.dll on steroids, having/planning almost full wpf features. Back in maybe 2006 I wished to have SL on industrial NETCF terminals, I hoped a lot, as SL is in fact, fast native engine writing directly to GPU/DX/something, I dont know, with NETCF scripting. Not having that, I made apps tageted to NETCF, using tiny subset of winforms where compiled assemblies worked on desktop for development, so I tweaked the VS tooling with some #ifdefs and lived with that, having tiny DI runtime assembly switching with pinvoked and glued things to custom device sdk and shared sqlservercompact etc... back to topic: what about best possible tooling around NeosisGUI "for now" (maybe some MS investment to these guys) ... I know its tiny and primarily targeted to games, but they made it runing everywhere, it seems
... no flames, peace, of course :)
https://channel9.msdn.com/blogs/tina/Silverlight-Rehab/

**Comment by mdtauk** [comment]
XAML seems pretty contained, and rather than replacing XAML or augmenting a new Hybrid XAML - any new solution should probably exist alongside XAML, and use the same Rendering stack.  A new parsing engine could be made.  

CSS for styling makes sense, as it is very familiar, and may just need a doctype equivalent to mark it as WinUI.  

As for the markup, or UI building documents - That will need a lot of thought...

* SwiftUI uses Swift, so for Windows, it would be non-compiled C#, VB or C++ (Maybe XLang solves this approach so you don't need to make 3, 4 etc different parsers.)  

* HTML is likely too primitive and fixed in its ways.  

* JSON would drive people mad, with how easy it is to malform.

 XAML has been around for over a decade, and has equivalents like XIB and AMXL to learn from.  Windows is now using XAML internally for Shell code, and so there should probably a heart to heart discussion, company wide:
- Where are the pain points with XAML?
- Where does XAML exhibit quirks which are vastly different to competing frameworks/platforms?
- Where is writing UI done better?
- How do we develop a new declarative UI pattern, with flexibility for future additions, and potential support for cross platform dev, Xamarin, React, etc.

**Comment by Odonno** [comment]
@mdtauk Answering this question is really hard. I think that to fully answer this question, someone need an holistic view of every UI tools/specification available in our hands with a large spectrum in order to bring the best vision around this subject.

IMO, I believe that the Windows ecosystem needs a declarative UI language and it should target only the Windows ecosystem to do one thing and doing it right, starting with the UWP for a start, why not? Because react does a great job but for the web only (react-native does not count here), jetpack compose for Android and SwiftUI for iOS.

Before going further, I need to tell you that I know HTML/CSS/JS combo and C#/XAML combo well. I am used to angular and react these days and I am in love with react & jsx. I also did some v-dom stuff which is kind of the root of react so I have a good understand of virtual-dom and the benefits/inconveniences. That's a great start but like I said, still need to get deeper knowledge with latest UI patterns (Flutter, Jetpack, SwiftUI).

So from my perspective, I think a declarative UI language can have a clear benefits over C#/XAML:

* XAML has a good data-binding but you often have to take care of stuff in your C# code (transforming your object). But not a big deal (you have converter right?).
* Event handling makes it hard to work every day. Ok, with MVVM pattern you have commands and you can pass a parameter. But then what about passing multiple params or a big data object with everything you need. And also, you cannot make a reference to a XAML element. Compare it with a js function in angular/react and you will know :)
* And what you will expect is components... Making tiny reusable components.. Both stateless and stateful components is really making devs productive. Of course, you can use `Style` property on some design elements in XAML and/or create custom elements but then you will have to deal with passing `props` down and, you know..

That's a lot of criticism but I assure you that I have nothing against XAML. I like the way it brings the whole vision of the view in one page and it's great. But when making huge apps or having some modularity in your app, you'd like to `compose` the UI and XAML does not offer that.. Well, that's an argument against inheritance but I suppose that's why the libraries we are talking are a trend.

Finally, I really love the new React hooks feature and it can be a bit scary but it is really smart how they did it. It works really great in their "engine" and I suppose you could make some if you were in a React/XAML world:

* `useState`/`useEffect` = the default ones
* `useTransition`/`useCompositionAnimation` for animations
* `useRouting` for navigation
* and so on..

Well, to have something more concrete about this imaginary language, I started to write something some months ago and I'be honest, it takes quite some time because

* you have to make sure the default system works (in pure C#) and that basic UI controls are integrated (containers, buttons, text elements, etc..)
* you need a new language like jsx, something like XAML-like syntax in C# and so you need a parsing tool and so you need some kind of compiler for this language (that compiles into pure C#)
* testing different language specification (having a concise declarative code and being able to extend the language)

But then, you have to remember the pain point is that a XAML tree is clearly not a DOM tree, which can lead to a more longer talk. :) To be short, I managed to make a virtual XAML tree by reusing existing elements even in extreme conditions where a virtual DOM would have rewrite elements. That kinda make me think it looks like a combination of both react and imba.

To conclude, that's still a PoC and I am far from a stable project but I get it that more and more people are waiting for such a thing. We now need to figure it out.

**Comment by mdtauk** [comment]
@Odonno  Xaml does have User Controls, and you can make Custom Controls - but these can be tricky to create, and use.  There is a lot to be said for simplifying the Tree.  Making each control or component a self contained element, which can be positioned anywhere.

**Comment by michael-hawker** [comment]
I feel like if things go down this route at all, it should be built on top of the ideas of **XAML Direct** and just the Native XAML code-behind instantiation.  Maybe some of the improvements to C# object instantiation in general can help in tandem with this type of work.

Having worked with various code based UI frameworks in the past in various languages, the power of XAML is really the ease in which one can copy and paste things around in the XML structure without having to re-work anything.  With code, there's a lot of other syntax that gets in the way when trying to move things around, especially when code is needed to direct the relations of UI elements to one another as well.

It seems like a large effort to create a new system, but if we can improve the existing systems to aid others with other UI paradigms in transitioning to leveraging XAML itself in combination, it could be useful to approach from that angle from a language and construction view in general.

**Comment by mdtauk** [comment]
@michael-hawker  Bearing in mind WinUI will need to also support C/C++, VB, and possibly JS with React Native.

**Comment by michael-hawker** [comment]
@mdtauk good point, I get hyper-focused sometimes on the .NET side of things from the Toolkit.  

Though I think that also shows where XAML itself can shine as it's a well understood and powerful abstraction that doesn't need to tie-itself to the language driving it. 😉

**Comment by mdtauk** [comment]
> @mdtauk good point, I get hyper-focused sometimes on the .NET side of things from the Toolkit.
> 
> Though I think that also shows where XAML itself can shine as it's a well understood and powerful abstraction that doesn't need to tie-itself to the language driving it. 😉

Its mainly the verbosity of XAML, as well as the rigid syntax, and no separation of Styling from UI Layout - that makes it awkward to work with.

**Comment by VincentH-Net** [comment]
@michael-hawker
> Having worked with various code based UI frameworks in the past in various languages, the power of XAML is really the ease in which one can copy and paste things around in the XML structure without having to re-work anything. With code, there's a lot of other syntax that gets in the way when trying to move things around, especially when code is needed to direct the relations of UI elements to one another as well.

Could you give/point to some example XAML and indicate a part that would be harder to move around if it were coded UI? I'm wondering if your experience is due to imperative coded UI versus declarative (like [CSharpForMarkup](https://github.com/VincentH-Net/CSharpForMarkup)). I'd like to try out an example and improve CSharpForMarkup if possible.

**Comment by Odonno** [comment]
@VincentH-Net I really like the approach you take with [CSharpForMarkup](https://github.com/VincentH-Net/CSharpForMarkup). 

I am currently working on a huge new version of [ReduxSimple](https://github.com/Odonno/ReduxSimple), with simplified reducers, sub-reducers, rewritten selectors, effects and routing. 

I would like to use your library in the samples of ReduxSimple because it can be a perfect match to combine a Redux architecture and a Functional UI framework. The thing is that I started my first sample app with UWP (before WPF and Xamarin) because I needed to start somewhere obviously. :) Do you think CSharpForMarkup can work with UWP?

**Comment by VincentH-Net** [comment]
@Odonno Glad to hear you like it!
> Do you think CSharpForMarkup can work with UWP?

The approach I took with [CSharpForMarkup](https://github.com/VincentH-Net/CSharpForMarkup) for Xamarin Forms can be applied to any .NET UI framework that uses data binding.

Since CSharpForMarkup is just a thin set of helper functions directly on top of the UI framework, you could effectively create a close lookalike of these helpers for UWP. Some helpers for Forms are specific but most will have a close equivalent on UWP.

You're welcome to do a PR to add a UWP port on CSharpForMarkup :-)
Simply copy [XamarinFormsMarkupExtensions.cs](https://github.com/VincentH-Net/CSharpForMarkup/blob/master/src/XamarinFormsMarkupExtensions.cs)
to UwpMarkupExtensions.cs and refactor it to UWP.

You might want to consider porting your UWP sample to UNO since that virtually is 1 on 1 compatible with UWP. Now that is a PR id _really_ love to see :-)

**Comment by Odonno** [comment]
@VincentH-Net Good to know. I'll give a look at it once the 3.0 of ReduxSimple is finished.

**Comment by charlesroddie** [comment]
UI defined in code can achieve the following now:
- Similar logical structure as XAML
- Type safety of setting properties, unlike XAML which uses various string names and conversions
- Type safety of bindings, by avoiding existing WPF/UWP/Xamarin untyped bindings
- No need to define bindable properties any more, only standard settable properties
- Binding sources can be much more flexible using a typed [reactive](https://en.wikipedia.org/wiki/Reactive_programming) approach
- Functions that are easy to defined and accessed as objects rather than by name lookup, and much easier to work with than bindings.

For example we can already write:
```fsharp
// in a ViewModel
let text = Mutable.create ""
let capitalizedText = text |> Signal.map (fun s -> s.ToUpper())
let add() = text.Value <- text.Value + "a"

// in a View
let label = Label()
capitalizedText.bind label.set_Text
StackLayout(
    Views = [
        label
        Button(Text = "Add", Click = add)
        ]
    )
```

Useful future tooling would include:
- A form of live views and/or hot reload, including a mechanism for sample data. For example:
```fsharp
type FooLabel(bar:string) =
    inherit Label(Text = "foo"+bar)
    [<Preview>] // only used for live preview tooling
    static member Sample = FooLabel("bar")
```

**Comment by surfsky** [comment]
I like QML & Razor synax:

``` js
@UserControl
@Class ControlCatalog.Pages.ButtonPage
@Using System.Drawing

StackPanel {
    Orientation:"Vertical" 
    Spacing:"4"
    TextBlock{Classes:"h1" Text:"Button"}
    TextBlock{Classes:"h2" Text:"A button control"}
    StackPanel{
        Orientation:"Horizontal"
        Margin:"0,16,0,0"
        HorizontalAlignment:"Center"
        Spacing:"16"
        StackPanel{
            Orientation:"Vertical" Spacing:"8" Width:"150"
            Button{Text: "Button"}
            Button{Foreground:"White" Text:"Foreground"}
            Button{Background:"@ThemeAccentBrush" Text:"Background"}
            Button{IsEnabled:"False" Text:"Disabled"}
            Button{
                Text:"Re-themed"
                Style.Resources {
                    SolidColorBrush {ThemeBorderMidBrush:"Red"}
                    SolidColorBrush {ThemeControlHighBrush:"DarkRed"}
                }
            }
        }
        StackPanel{
            Orientation:"Vertical" Spacing:"8" Width:"150"
            Button {BorderThickness:"0" Text:"No Border"}
            Button {BorderBrush:"@ThemeAccentBrush" Text:"Border Color"}
            Button {BorderBrush:"@ThemeAccentBrush" BorderThickness:"4" Text:"Thick Border"}
            Button {BorderBrush:"@ThemeAccentBrush" BorderThickness:"4" IsEnabled:"False" Text:"Disabled"}
        }
    }   
}
```

And QML binding syntax:

```
SliderBar { Id=“sld” Min=0 Max=100}
TextBlock { Text: sld.Value }
```

The grammar of QML is well refined, just liking poem. QML has several advantages:

1. Syntax like JSON
2. Concise binding syntax
3. Concise property syntax
4. Signal and slot
5. ...

The first 3 items can be adopted.
By the way, the file extension may be ".csml", It means "csharp markup language". 
:)

**Comment by gulshan** [comment]
I think the specialty of UI frameworks like SwiftUI and Jetpack Compose is putting logical operations like if conditions and loops while defining UI code. Swift and Kotlin has some language features, that makes this possible. This will be hard to do in C#.

**Comment by VincentH-Net** [comment]
@gulshan said:
> I think the specialty of UI frameworks like SwiftUI and Jetpack Compose is putting logical operations like if conditions and loops while defining UI code. Swift and Kotlin has some language features, that makes this possible. This will be hard to do in C#.

Actually, it's not that hard. #CSharpForMarkup is now built-in in Xamarin Forms as the **C# Markup** feature, and makes it easy to combine declarative and procedural buildup of markup.
It is currently in the 4.6 prerelease NuGet. The documentation will go live when 4.6 goes stable; for now see [the merged Xamarin Forms PR](https://github.com/xamarin/Xamarin.Forms/pull/8342) for full documentation.

@surfsky In contrast to Razor-like approach C# Markup does not add / mix new language elements to C#, and avoids strings for data binding or property values, choosing typesafe, intellisense-supported C# features instead. However the markup structure reads similar to your example.

**Comment by gulshan** [comment]
@VincentH-Net I followed the PR and I really admire what you did. But what I tried say is, Jetpack compose supports code like this-
```kotlin
Column {
    people?.forEach { person ->
        Padding(16.dp) {
            Text(text = "${person.name} (${person.craft})")
        }
    }
}
```
We cannot put a loop or a conditional statement within a collection initializer in C#.

**Comment by VincentH-Net** [comment]
@gulshan said
> I followed the PR and I really admire what you did. ... 
> We cannot put a loop or a conditional statement within a collection initializer in C#

Thanks for the complement!

Actually you **can** do this because #CSharpForMarkup has built-in helpers to mix-in logic with declarative markup: `.Invoke()` and `.Assign()`.
E.g. above could be:
```CSharp
new StackLayout { Spacing = 16 }
    .Invoke(s => people?.ForEach(person => s.Children.Add(
        new Label { Text = $"{person.Name} ({person.Craft})" }
    )
))
```

Or, if you want the label text to update to changes in person, you can add a calculated property and bind it like this:
```CSharp
class Person : BaseViewModel
{
    public string Name { get; set; }
    public string Craft { get; set; }

    public string NameAndCraft => $"{Name} ({Craft})}";
}

new StackLayout { Spacing = 16 }
    .Invoke(s => people?.ForEach(person => s.Children.Add(
        new Label { } .Bind (nameof(Person.NameAndCraft))
    )
))
```

Or, you could factor out the logic to a method that either returns a new `StackLayout` or takes the `StackLayout.Children` as parameter.

Or, if you want to support hot reload optimally, you could `.Assign (out peopleStack)` and at the end of the Build() method call a `void BuildPeopleStack()` method that sets `peopleStack.Children`

So there are quite a few options to do this, both inline and in separate methods. All because the versatility of C#. No magic, new syntax or strings needed :-)

**Comment by gulshan** [comment]
@VincentH-Net Actually SwiftUI and Jetpack Compose both use a little bit of code generation along with built-in language syntax, specially for "Container" view (with children views), like this- https://stackoverflow.com/a/56474164/97714
I don't know if this is possible in C#. I think it will be something really nice to have in C#.

**Comment by luismts** [comment]
I think @VincentH-Net  the challenge now is getting C # HotReload into VS. If this is achieved, the number of people using C # projects will be massive.

LiveSharp is great, and it's not expensive. But many do not know it, and it is a disadvantage. The simple fact of knowing that you have to install a third-party tool, and more being paid, leaves this option for a specific audience.

**Comment by weitzhandler** [comment]
@luismts I shall say hot reload + hot restart.

**Comment by JaggerJo** [comment]
With a language that's more suitable for DSLs the UI declaration can be just code. 

We don't need yet another Razor/xaml thing IMHO..

The code below is a sample of (https://github.com/AvaloniaCommunity/Avalonia.FuncUI) with [F#](https://fsharp.org). Might be interesting... here are [templates](https://github.com/AvaloniaCommunity/Avalonia.FuncUI.ProjectTemplates) if you want to give it a spin.

```fsharp
module Counter =

    type CounterState = {
        count : int
    }

    let init = {
        count = 0
    }

    type Msg =
    | Increment
    | Decrement

    let update (msg: Msg) (state: CounterState) : CounterState =
        match msg with
        | Increment -> { state with count =  state.count + 1 }
        | Decrement -> { state with count =  state.count - 1 }
    
    let view (state: CounterState) (dispatch): IView =
        DockPanel.create [
            DockPanel.children [
                Button.create [
                    Button.onClick (fun _ -> dispatch Increment)
                    Button.content "click to increment"
                ]
                Button.create [
                    Button.onClick (fun _ -> dispatch Decrement)
                    Button.content "click to decrement" 
                ]
                TextBlock.create [
                    TextBlock.dock Dock.Top
                    TextBlock.text (sprintf "the count is %i" state.count)
                ]
            ]
        ]    
```

**Comment by asklar** [comment]
Shameless plug, check out a reactive extension to C#/XAML that I wrote for last year's hackathon, called CSX. It lets you have a syntax like JSX suck as:

```jsx
<StackPanel
    Id="taskList"
    Grid.Row="1"
    Grid.Column="0"
    Grid.ColumnSpan="2">
    {
        return person.Tasks.Select(todo => CreateToDoItem(todo));
    }
</StackPanel>
<Grid.PointerReleased Handler={
        completionRateTextBlock.Opacity = 1.0;
        taskList.Visibility = taskList.Visibility == Visibility.Collapsed ? 
                              Visibility.Visible : Visibility.Collapsed;
    }/>
```

https://github.com/asklar/CSX

**Comment by mdtauk** [comment]
Xamarin is working on what it calls **C# Markup**

https://docs.microsoft.com/en-us/xamarin/xamarin-forms/user-interface/csharp-markup

**Comment by VincentH-Net** [comment]
See [this spec](https://github.com/dotnet/maui/issues/119) for the next version of **C# Markup** in  [.NET MAUI](https://devblogs.microsoft.com/dotnet/introducing-net-multi-platform-app-ui/) (top: MVU, bottom: MVVM):
![image](https://user-images.githubusercontent.com/1872271/84620619-c6d15d00-aed8-11ea-9af1-44830cb84787.png)

**Comment by Odonno** [comment]
@VincentH-Net That looks really interesting.

**Comment by sonnemaf** [comment]
I have created an [proof of concept](https://github.com/sonnemaf/MvuTest) for a MVU pattern in UWP. Maybe this can extend this discussion.

```
using MvuTest.Controls;
using System;
using Windows.UI;
using Windows.UI.Popups;
using Windows.UI.Xaml;

namespace MvuTest {

    public class MainPage : MvuPage {

        private TextBlock2 _tb;

        private int _count;

        protected override IUIElement2 Build() =>
            this.StackPanel(
                this.ToggleButton("Toggle it")
                    .Width(200)
                    .FontSize(30),
                this.Button("Increment", (sender, e) => {
                    _tb.Text($"Count {++_count}");
                }).HorizontalAlignment(HorizontalAlignment.Center),
                (_tb = this.TextBlock("Count 0"))
                    .FontSize(20)
                    .Foreground(Colors.Red),
                this.TextBox("Fons Sonnemans")
                    .Header("Name")
            ).Spacing(12)
             .HorizontalAlignment(HorizontalAlignment.Center)
             .VerticalAlignment(VerticalAlignment.Center);
    }

}
```

**Comment by mdtauk** [comment]
What are the chances of getting **C# Markup** as a first class option for WinUI development?

Looking at the latest Google IO and WWDC events, we now see a trend which follows on from Flutter and React Native for **Declarative UI**

Swift and SwiftUI for macOS and iOS

Kotlin and Jetpack Compose for Android

Will a decision be made for WinUI to _prefer_ C# and C# Markup - so push resources to it, feature it over Xaml in sample apps, templates, conference demos etc.

This also brings up some issues.  WinUI and Reunion supports C++ and various other language abstractions - so Xaml will need to remain for those other languages right?

**Comment by sonnemaf** [comment]
I have the feeling that this is not high on the priority list. I don't think that this is really bad. I love XAML. It is great, consistent and not very difficult. It is verbose though. I would love to have a C# markup solution. But I don't need it. It would be a bonus. I think that other WinUI features like input validation are more essential for its success.

**Comment by mdtauk** [comment]
> I have the feeling that this is not high on the priority list. I don't think that this is really bad. I love XAML. It is great, consistent and not very difficult. It is verbose though. I would love to have a C# markup solution. But I don't need it. It would be a bonus. I think that other WinUI features like input validation are more essential for its success.

Sure it's not a priority in terms of functionality, but Microsoft is missing a shift in developers expectations, as every other platform and framework is moving this way.

**Meeting developers where they are** this is Microsoft's current mantra

**Comment by Happypig375** [comment]
As I said, just use C#. C# provides a concise way to specify views already.

**Comment by RichiCoder1** [comment]
> As I said, just use C#. C# provides a concise way to specify views already.

Compared to some of the tools mentioned above, it's not quite nearly as concise. Various libs and tools have made it better but not nearly as pleasant. This discussion item outlines it well: https://github.com/dotnet/maui/discussions/119#discussioncomment-59504

**Comment by VincentH-Net** [comment]
@mdtauk I have been championing C# Markup for all .NET UI frameworks for a couple of years now, and built a Gen 1 for Xamarin Forms (in XCT now). I pushed for proper C# hot reload which is now coming. I also have a nearly finished next-gen C# Markup version for WinUI 3 and UNO on the shelf. A small snippet ([source](https://twitter.com/vincenth_net/status/1341023880249532420)):
![image](https://user-images.githubusercontent.com/1872271/121731587-71328f00-caf1-11eb-8b83-cccf4e58c594.png)

However, to deliver C# Markup in a way so it can compete with Flutter and SwiftUI, also some tooling is needed: [proposal](https://twitter.com/vincenth_net/status/1342120855879479297)

This requires assigning / hiring at least one high level engineer to implement and support this - for one or more .NET UI frameworks.

The past 6 months I am just waiting for any of the UNO / WinUI / MAUI teams - or maybe a team that targets multiple .NET UI frameworks, like the XAML tooling team - to wake up to the outside world and come to the realization that using a single language for declarative markup and logic is not the future but the present - and has been for a while now.

All arguments were already clearly discussed with involved teams (most with MAUI and UNO, WinUI 3 team not so much yet - they reacted a bit confused when I mentioned C# markup in their community calls but they indicated they had too much work and I could add it myself after they go open source).

So, the waiting is for the moment when MS / UNO think this is worth at least a single engineer's time (note that MAUI MVU is also just a single person, who has other tasks as well). If they want to hire me I might even say yes ;-)

**Comment by mdtauk** [comment]
@VincentH-Net  I agree completely.  It needs someone within Microsoft to decide it needs to be built into the tooling, and given priority.

People adopting MAUI will be a big push, as I suspect that team will extol the virtues of a move away from XAML files.

But MAUI is C#, and whilst C# Markup makes total sense there, it doesn't cover C++ development.

MAUI will also be a subset of WinUI's feature set, and so will be limiting in that way.

The removal of the XAML designer is another thing to hamper Xaml dev.
